File: Algorithms/Python/42 Trapping Rain Water.py
# idea: find the max value, then divide the list from the max value index into two: left and right.
# The max area must be the sum of max left area and max right area.
# Notice the boundary condition.
class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) == 0 or max(height) == 0:
            return 0

        max_index = height.index(max(height))
        if max_index == 0:
            return self.get_right_area(max_index, len(height) - 1, height)
        if max_index == len(height) - 1:
            return self.get_left_area(0, max_index, height)

        return self.get_left_area(0, max_index, height) + self.get_right_area(max_index, len(height) - 1, height)
    
    def get_left_area(self, left, right, height):
        # height[right] is the max value
        if left >= right - 1:
            return 0

        # find the max value between [left, right - 1]
        if max(height[left: right]) == 0:
            return 0

        max_index = height[left: right].index(max(height[left: right])) + left
        # area between [max_index, right]
        area = height[max_index] * (right - max_index - 1) - sum(height[max_index + 1: right])
        return area + self.get_left_area(left, max_index, height)

    def get_right_area(self, left, right, height):
        # height[left] is the max value
        if left >= right - 1:
            return 0

        # find the max value between [left + 1, right]
        if max(height[left + 1: right + 1]) == 0:
            return 0
        
        max_index = height[left + 1: right + 1].index(max(height[left + 1: right + 1])) + (left + 1)
        # area between [left, max_index]
        area = height[max_index] * (max_index - left - 1) - sum(height[left + 1: max_index])
        return area + self.get_right_area(max_index, right, height)       

File: Algorithms/C/42 Trapping Rain Water.c
int index_of_max_value(int left, int right, int* height)
{
    int max_value_index = left;
    for (int i = left + 1; i <= right; ++i){
        if (height[i] > height[max_value_index]){
            max_value_index = i;
        }
    }
    return max_value_index;
}

int sum(int left, int right, int* height)
{
    int s = 0;
    for (int i = left; i <= right; ++i){
        s += height[i];
    }
    return s;
}

// assume height[right] is the max value between [left, right]
int get_left_area(int left, int right, int* height)
{
    if (left >= right - 1){
        return 0;
    }

    // find the max value between [left, right - 1]
    int max_value_index = index_of_max_value(left, right - 1, height);
    if (height[max_value_index] == 0){
        return 0;
    }

    // caculate the area between [max_value_index + 1, right - 1]
    int area = height[max_value_index] * (right - 1 - (max_value_index + 1) + 1) - sum(max_value_index + 1, right - 1, height);
    return area + get_left_area(left, max_value_index, height);
}

// assume height[left] is the max value between [left, right]
int get_right_area(int left, int right, int* height)
{
    if (left >= right - 1){
        return 0;
    }

    // find the max value between [left + 1, right]
    int max_value_index = index_of_max_value(left + 1, right, height);
    if (height[max_value_index] == 0){
        return 0;
    }

    // caculate the area between [left + 1, max_value_index - 1]
    int area = height[max_value_index] * (max_value_index - 1 - (left + 1) + 1) - sum(left + 1, max_value_index - 1, height);
    return area + get_right_area(max_value_index, right, height);
}

int trap(int* height, int heightSize){
    if (height == NULL || heightSize == 0){
        return 0;
    }

    int max_value_index = index_of_max_value(0, heightSize - 1, height);
    if (height[max_value_index] == 0){
        return 0;
    }
    if (max_value_index == 0){
        return get_right_area(max_value_index, heightSize - 1, height);
    }
    if (max_value_index == heightSize - 1){
        return get_left_area(0, max_value_index, height);
    }
    return get_left_area(0, max_value_index, height) + get_right_area(max_value_index, heightSize - 1, height);
}
